OttoWinter/esphomeyaml
# Read/Write bootloader, partition table and any partition via OTA

# What does this implement/fix?

This adds support to update the bootloader and partition tables with OTA. It allows migrating from
firmwares with different layouts (new Tasmota versions) or repair botloaders (such as the locked new Shellies).

Updating the bootloader and partition tables only works with the esp-idf framework. For arduino [arduino-esp32](https://github.com/espressif/arduino-esp32/releases/tag/3.0.0-alpha3) is needed. This is currently in alpha so no platformio package is available. It is possible to move from esp-idf to arduino OTA so you can migrate Tasmota -> esphome(esp-idf) -> esphome(arduino).

## New CLI commands

```bash
# update OTA the bootloader without rebooting
python -m esphome -v upload test.yaml --bootloader --no-reboot
# update OTA the partition table without rebooting (requires esp32-idf version > 5.1)
python -m esphome -v upload test.yaml --partition-table --no-reboot
# update OTA partition with type and subtype
python -m esphome -v upload --partition-type=0x0 --partition-subtype=0x10 test.yaml
# update OTA partition with label
python -m esphome -v upload --partition-label=app0 test.yaml
# download OTA the bootloader
python -m esphome -v download test.yaml --bootloader --file=bootloader.bin
# download OTA the partition table without rebooting (requires esp32-idf version > 5.1)
python -m esphome -v download test.yaml --partition-table --file=partition_table.bin
# download OTA partition with type and subtype
python -m esphome -v download --partition-type=0x0 --partition-subtype=0x10 test.yaml  --file=firmware.bin
# download OTA partition with label
python -m esphome -v download --partition-label=app0 test.yaml  --file=firmware.bin
# update OTA all esphome parts (equivalent to a factory reset)
python -m esphome upload-factory-ota test.yaml
```

The `upload-factory-ota` command makes a backup of the partition table before flashing the new one. If any operation fails then the backup is automatically restored. This usually happens when the new firmware would override the running partition.

## Tasmota migration

If you are migrating from Tasmota 1.12 or newer using the Safeboot partition you need the following steps. There are 2 options depending on the used esp-idf framework.

### Single command (requires esp-idf>5.1)

This is how to do it without custom partition_tables.

1. Build an ESPHome (`test.yaml`) with a minimal configuration (with wifi and OTA) with esp-idf>5.1
```yaml
# Example for esp32-c3 (Shelly Plus 1PM Mini)
esphome:
  name: test
  platformio_options:
    board_build.flash_mode: dio
esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    platform_version: 6.4.0
    version: 5.1.1
  variant: esp32c3
wifi:
  manual_ip:
    static_ip: 192.168.0.33
    gateway: 192.168.2.1
    subnet: 255.255.192.0
    dns1: 192.168.2.1
  networks:
  - ssid: my_wifi
    password: my_wifi_password
ota:
  password: my_ota_password
  unprotected_writes: True # This is mandatory if you want to flash the partition table or bootloader!
logger:
  hardware_uart: USB_SERIAL_JTAG
web_server:
button:
- platform: restart
  name: Restart
```
2. Use Tasmota firmware upgrade with the ESPHome legacy firmware binary built in the previous step. ESPHome should boot
3. Upgrade all ESPHome to factory: `python -m esphome -v upload-factory-ota test.yaml`. ESPHome should reboot and use ESPHome partition table and boot loader.

Afterwards you might downgrade to the recommend esp-idf framework.

### Using multiple commands

If you are using the recommend esp-idf framework you will need to use a custom partition table:

1. Build an ESPHome (`test.yaml`) with a minimal configuration (with wifi and OTA) and the following partition table (this uses the Tasmota OTA data partition and app1 address/size):
```CSV
# Name,   Type, SubType, Offset,   Size, Flags
otadata,  data, ota,     0x00e000,        0x2000,
# phy_init, data, phy,     ,        0x1000,
app0,     app,  ota_1,   0x010000,      0x0d0000,
app1,     app,  ota_0,   0x0e0000,      0x0d0000,
nvs,      data, nvs,     0x38C000,       0x6d000,
```
```yaml
# Example for esp32-c3 (Shelly Plus 1PM Mini)
esphome:
  name: test
  platformio_options:
    board_build.partitions: "../../../custom_partitions.csv"
    board_build.flash_mode: dio
esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
  variant: esp32c3
wifi:
  manual_ip:
    static_ip: 192.168.0.33
    gateway: 192.168.2.1
    subnet: 255.255.192.0
    dns1: 192.168.2.1
  networks:
  - ssid: my_wifi
    password: my_wifi_password
ota:
  password: my_ota_password
  unprotected_writes: True # This is mandatory if you want to flash the partition table or bootloader!
logger:
  hardware_uart: USB_SERIAL_JTAG
web_server:
button:
- platform: restart
  name: Restart
```
2. Use Tasmota firmware upgrade with the ESPHome legacy firmware binary built in the previous step. ESPHome should boot
3. Upgrade the partition table: `python -m esphome -v upload test.yaml --partition-table`. ESPHome should reboot and still run from address 0x0e0000 as in Tasmota
4. Upgrade the ESPHome (either using the dashboard or `python -m esphome -v upload test.yaml`). ESPHome should boot and it will be running out of the 0x010000 address as in regular ESPHome. At this point OTA works but in order to use all available flash we need the following steps.
5. Build ESPHome (`test.yaml`) without the custom partition table from step 1
6. Upgrade the partition table: `python -m esphome -v upload test.yaml --partition-table`


## Types of changes


**Related issue or feature (if applicable):** NA
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:


```yaml
# Example config.yaml
ota:
  unprotected_writes: True # This is mandatory if you want to flash the partition table or bootloader!
  password: removedPassword
```

## Checklist:
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).
